### ISSUE TITLE
map iterator - removing element while iteration

### ISSUE THREAD

**Issue by Yaffle** [eriwen/es6-map-shim#1]
var map = new Map();
map.set(1, 1);
map.set(2, 2);
map.set(3, 3);

var iterator = map.iterator();
console.log(iterator.next()[0]);
map["delete"](1);
console.log(iterator.next()[0]);

**Comment by eriwen** [comment]
@Yaffle I don't see anything in the spec that specifically states what should happen in this case, but as you wrote it, the console would print `1` and then `3`. 

What _should_ happen?

**Comment by Yaffle** [comment]
yes, spec is not good defined here.... seems, Map does not support modification while iteration in the spec, but Firefox implements it

``` javascript

var map = new Map();
map.set(1, 1);
map.set(2, 2);
map.set(3, 3);

var iterator = map.iterator();
console.log(iterator.next()[0]);
map["delete"](1);
console.log(iterator.next()[0]);
console.log(iterator.next()[0]);
```

this codes outputs "1\n2\n3\n" in the console in FF

**Comment by eriwen** [comment]
I hesitate to follow just one browser vendor's implementation without support from the spec. 

I think it is best to wait and see how it's implemented in at least one other major browser before this project follows suit. Chrome Canary's experimental impl does not yet include `.iterator()`. I'll update this issue when it does.

**Comment by Yaffle** [comment]
it is interesting how to impelement this....

Java LinkedHashMap just throws exception in this case - http://docs.oracle.com/javase/1.4.2/docs/api/java/util/LinkedHashMap.html

> if the map is structurally modified at any time after the iterator is created, in any way except through the iterator's own remove method, the iterator will throw a ConcurrentModificationException.

**Comment by medikoo** [comment]
From ES6 latest draft (15.14.4.4):

> forEach does not directly mutate the object on which it is called but the object may be mutated by the calls to callbackfn.
> NOTE Each key is visited only once with the value that is current at the time of the visit. If the value associated with a key is modified after it has been visited, it is not re-visited. Keys that are deleted after the call to forEach begins and before being visited are not visited. New keys added, after the call to forEach begins are visited.

.. and as I checked with latest Firefox, iterators follow exactly same logic.

**Comment by Yaffle** [comment]
> I hesitate to follow just one browser vendor's implementation without support from the spec.
> I think it is best to wait and see how it's implemented in at least one other major browser before this project follows suit. Chrome Canary's experimental impl does not yet include .iterator(). I'll update this issue when it does.

@eriwen, it is in the spec and it is in two browsers (IE 11 & Firefox)

**Comment by Yaffle** [comment]
update for test (IE 11 supports only `forEach`): 

``` javascript
(function () {
  var s = [];
  var value = 42;
  var map = new Map();
  map.set(0, value);
  map.set(1, value);
  map.set(2, value);
  map.set(3, value);
  map.delete(1);
  map.forEach(function (value, key) {
    s.push(key);
    if (key === 0) {
      map.delete(0);
      map.delete(2);
      map.set(4);
    }
  });
  alert(s.join(""));//should alert 034
}(self));
```
